framework/include/samplers/Sampler.h

```diff
@@ -15,71 +15,102 @@
 #ifndef SAMPLER_H
 #define SAMPLER_H
 
+// libMesh includes
+#include "libmesh/dense_matrix.h"
+
+// MOOSE includes
 #include "MooseObject.h"
-#include "Distribution.h"
-#include "Restartable.h"
+#include "MooseRandom.h"
+#include "SetupInterface.h"
 #include "DistributionInterface.h"
-#include "RandomInterface.h"
+#include "Distribution.h"
 
 class Sampler;
 
 template <>
 InputParameters validParams<Sampler>();
 
 /**
- * This is the base class for Samplers and all Samplers should inherit from this class
+ * This is the base class for Samplers.
+ *
+ * A sampler is responsible for sampling distributions and providing an interface for providing
+ * the samples to objects. The sampler is designed to handle any number of random number generators.
+ *
+ * The main method in this object is the getSamples method which returns the distribution samples.
+ * This method will return the same results for each call when the "_generator" from this class
+ * is used by child classes.
+ *
+ * Samplers support the use of "execute_on", which when called results in new set of random numbers,
+ * thus after execute() runs the getSamples() method will now produces a new set of random numbers
+ * from calls prior to the execute() call.
  */
-class Sampler : public MooseObject,
-                public RandomInterface,
-                public DistributionInterface,
-                public Restartable
+class Sampler : public MooseObject, public SetupInterface, public DistributionInterface
 {
 public:
   Sampler(const InputParameters & parameters);
+
   /**
-   * Draw the values from given distributions for all the perturbed parameters
+   * Return the list of distribution names.
    */
-  virtual void generateSamples();
+  const std::vector<DistributionName> & getDistributionNames() const { return _distribution_names; }
+
+  ///@{
   /**
-   * Return a vector of variable names
+   * Setup method called prior and after looping through distributions.
    */
-  virtual std::vector<std::string> getSampledVariableNames();
+  virtual void sampleSetUp(){};
+  virtual void sampleTearDown(){};
+  ///@}
+
+  /**
+   * Return the sampled distribution data.
+   */
+  std::vector<DenseMatrix<Real>> getSamples();
+
   /**
-   * Return a vector of current sampled values of given parameters
+   * Store the state of the MooseRandom generator so that new calls to getSamples will create
+   * new numbers.
    */
-  virtual std::vector<Real> getSampledValues(const std::vector<std::string> & variableNames);
+  void execute();
```

Comment:
The comment is specific, because that is what the method does and it isn't supposed to do anything else.